Extend CI Workflow to test multiple RMW

ruffsl opened the pull request:
This PR extends the CircleCI workflow for testing multiple RMW implementations. To compensate for the added build time in generating messages, a combination of ccache and CircleCI caches are used to accelerate fresh PR jobs, including jobs across successive workflows. To compensate for the added test time, RMW implementations is tested across parallel jobs using the quicker release build. 

---

## Motivation

In preparation for the security workshop at ROSCon, @mikaelarguedas and I have been working to demo SROS2 running on the Turtlebot3 to secure the entire ROS stack, including the navigation2. However, when using other Tier 1 supported RMW implementations to verify security functionality, we’ve been encountering a number of issues, including:

https://github.com/ros-planning/navigation2/issues/915
https://github.com/ros-planning/navigation2/issues/916

To help navigation team detect and debug these issues, I’d like to extend the current CI workflow to include support for testing multiple RMW implementations, as well as a number of other optimizations that may be helpful in keeping the CI tractable.

## Changes

* DockerHub and Dockerfile
  * Clarify underlay and overlay terminology and inconsistencies
  * Parameterize the `FROM` image using build a `ARG`
  * Add hooks for reusing Dockerfile for multiple tags
  * Install and use ccache for CI optimization

* CircleCI config.yaml
  * Clarify underlay and overlay terminology and inconsistencies
  * Split RMW testing for release builds into parallel jobs
  * Add nonce to cache keys to avoid collisions between parallel jobs and workflows
  * Use CI caching to preserve ccache across jobs for both underlay and overlay

<!--
### Before

![image](https://user-images.githubusercontent.com/2293573/62171013-57637680-b2e2-11e9-9ceb-a3b233c7788a.png)
-->

### After

![image](https://user-images.githubusercontent.com/2293573/62171073-98f42180-b2e2-11e9-84f5-6c7db24d6be7.png)


## Action Items

- [x] Update DockerHub automated build config
- [x] Change over image repo from ruffsl to rosplanning

![image](https://user-images.githubusercontent.com/2293573/62169463-20d72d00-b2dd-11e9-96c7-a9e1783673fc.png)

| Source Type | Source | Docker Tag | Dockerfile location                | Build Context   | Autobuild |
|-------------|--------|------------|------------------------------------|-----------------|-----------|
| Branch      | master | master     | /.dockerhub/nightly/Dockerfile     | <leave_empty>   | True      |
| Branch      | master | master-rmw | /.dockerhub/nightly-rmw/Dockerfile | <leave_ empty > | True      |

mkhansen-intel commented:
Can you help do the same with the Travis CI also? I'd like that to also run the tests with different RMW implementations.

SteveMacenski commented:
@mkhansen-intel I forget: why do we still use Travis if we have Circle + ROS PR builder going? I'm not sure what that adds the other 2 dont do

ruffsl commented:
> Can you help do the same with the Travis CI also?

Could we tackle this in a future/separate PR?

crdelsey commented:
Just FYI: opensplice middleware is broken too. #974.

crdelsey commented:
> why do we still use Travis if we have Circle + ROS PR builder going

Sometimes Circle will just mysteriously fail. Travis kind of acts like the backup when that happens. 
A couple of times now, Circle will aways fail when the gazebo_ros plugin is loaded. It fails with a symbol not found error as if the gazebo_ros plugin is linked with a different version of ROS than is currently present on the system.

After a week or two, it suddenly starts to work again. This has happened twice now; at least that I've noticed.

Also, due to the way the Circle build works, if master gets broken by an upstream change, a PR that fixes the problem can't make the CIrcle build pass. But it can make the Travis build pass.

I'm OK with not making Travis do all the builds Circle does, but, so long as it is isn't a huge burden, it is nice to have it as a backup check.

ruffsl commented:
> Circle will aways fail when the gazebo_ros plugin is loaded. It fails with a symbol not found error as if the gazebo_ros plugin is linked with a different version of ROS than is currently present on the system.

That sounds like an improper occurance of a cache restore loading build artifacts from a previous job. I've been using timestamps from the nightly docker image to keep the cache keys in lock step.

https://github.com/ros-planning/navigation2/blob/2350778c3eb11196de68ce006d5ec246e848f27b/.circleci/config.yml#L7-L9

In reworking the config for this PR, I realized that the cache keys between release and debug jobs (that run in parallel) could potentially collide, so in 54565dddd17b024ec4830df9934b757ddd20019a I added the `CIRCLE_CACHE_NONCE` env to help seed unique checksum.txt digests between jobs in the same workflow, or jobs with different compiler flags.

@crdelsey , if you could by chance remember an issue where this occurred, then I could dig into the respective CircleCI reports to assert if my hypothesis was the case, and if this PR resolves it.

ruffsl commented:
> Also, due to the way the Circle build works, if master gets broken by an upstream change, a PR that fixes the problem can't make the CIrcle build pass. But it can make the Travis build pass.

IIRC, master branch was broken for so long that `rosplanning/navigation2:master` had grown too stale, and fell out of sync from upstream changes that respective fixing PRs were expecting.

I was thinking we could have a CI Dockerfile that only install underlay and overlay dependencies, and skipps the colcon build layers, but then we'd miss out the benefit of having the DockerHub initialize our ccache each new PR.

Instead, I've added 664ec03 to resolve this *stalemate* (haha) buy having the dockerhub build hooks unset the `FAIL_ON_BUILD_FAILURE`, thus allowing the images to continue to build even when master breaks. However, the Dockerhub CI badge will be less indicative of colcon build success, so I've enables the nightly cron jobs in b40cdaa to keep a close eye. It should be less resource intensive now that we are using ccache.

ruffsl commented:
> Circle will aways fail when the gazebo_ros plugin is loaded. It fails with a symbol not found error as if the gazebo_ros plugin is linked with a different version of ROS than is currently present on the system.

I just also noticed that CircleCI branch name substitution using {{ .Branch }} does not prefix names with the repo name. Thus PRs from separate forks using the same branch name could collide.

To avoid cache collisions for PRs with same branch name, I've included the unique PR number into cache key. See b99b98a.

ruffsl commented:
I've also removed sharing of underlay caches across branches/PRs in 823173f.
That was probably a bug.

crdelsey commented:
@ruffsl Fortunately the latest Circle symbol failure problem was recent. Here is a build that shows the failure mode. 
https://circleci.com/gh/ros-planning/navigation2/1618?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

ruffsl commented:
> Here is a build that shows the failure mode.

Ok, looks like build [1618](https://circleci.com/gh/ros-planning/navigation2/1618) restored its underlay cache from build [1435](https://circleci.com/gh/ros-planning/navigation2/1435):

```
Found a cache from build 1435 at upstream-cache-v1-arch1-linux-amd64-6_85-7fIRLjO_6tpk1hegvoBhjHu+EAoQ4oDNc+JQ_VwWOlQ=
Size: 35 MB
Cached paths:
  * /opt/ros_ws
```

The the cache and originating build 1435 is comparatively old:
June 19 vs June 28.

```
Skipping cache generation, cache already exists for key: upstream-cache-v1-arch1-linux-amd64-6_85-7fIRLjO_6tpk1hegvoBhjHu+EAoQ4oDNc+JQ_VwWOlQ=
Found one created at 2019-06-19 20:58:00 +0000 UTC
```

Re-reading the latest docs, save_cache is a bit unitutive in that it [can't update existing caches](https://discuss.circleci.com/t/add-mechanism-to-update-existing-cache-key/9014/10).
Looks like the docs have been updated to better reflect this. Grrr... 

> Note If the cache for the given key already exists it won’t be modified, and job execution will proceed to the next step.
> https://circleci.com/docs/2.0/configuration-reference/#save_cache

> CircleCI restores caches in the order of keys listed in the restore_cache step. Each cache key is namespaced to the project, and retrieval is prefix-matched. The cache will be restored from the first matching key. If there are multiple matches, the most recently generated cache will be used.
> https://circleci.com/docs/2.0/caching/#restoring-cache

However, given that cache key matching is prefix based, I've reworked this in 479aca75d1b6054e67588ab1297fdcf12da9a7fd .
Some working examples jobs below, with 261 restored from 255, and 265 restored from 261:
* https://circleci.com/gh/ruffsl/navigation2/255
* https://circleci.com/gh/ruffsl/navigation2/261
* https://circleci.com/gh/ruffsl/navigation2/265

We shouldn't have another key unintended cache crosstalk for for about 267.484 years. :calendar: :crossed_fingers:

ruffsl commented:
Ok, so the `osrf/ros2:nightly-rmw` tags should be operational. I've merged origin master and my rmw branch on my fork to test out the CI on the latest state of this repo. Could I get an opinion on the test failures here, none of the jobs seem to be passing with master:

* debug: https://circleci.com/gh/ruffsl/navigation2/341
* release: https://circleci.com/gh/ruffsl/navigation2/342
  * test_rmw_opensplice_cpp: https://circleci.com/gh/ruffsl/navigation2/343
  * test_rmw_fastrtps_cpp: https://circleci.com/gh/ruffsl/navigation2/344
  * test_rmw_connext_cpp: https://circleci.com/gh/ruffsl/navigation2/345

ruffsl commented:
Since my last comment, changes I've added include:

- Add ccache stats and logs for debugging
- Split all build and test steps into separate jobs
- Use more yaml magic for keeping DRY
- switch to use official lcov orb
- enable test results
- enable parallel testing and load balancing

Example: https://circleci.com/gh/ruffsl/navigation2/418

I think this is ready for a review.

ruffsl commented on .circleci/config.yml:
```diff
@@ -1,240 +1,396 @@
+  #     name: Test Overlay
+  #     command: |
+  #       . install/setup.sh
+  #       src/navigation2/tools/run_test_suite.bash
```
Perhaps we could push any system integration tests into proper ctests  that include timing info in the xml test results, so that they can be load balanced accordingly. We could also split them out into their own job, but I'd like to make sure they are still reporting their own timing.

crdelsey replied:
> Perhaps we could push any system integration tests into proper ctests

They are are ctests already. Do you mean launch them from `colcon test`?

ruffsl replied:
> Do you mean launch them from colcon test

Yea, as then they could be load balanced by the circleci timing split. However splitting currently is at the granularity of package, so if one package take the lion share of test time, say more than the sum of the other tests; one unlucky worker will still end up testing that package by its lonesome.

crdelsey replied:
Ok. I don't want to go back to colcon running the system tests directly without figuring out some changes. I was unable to get the full console output when colcon ran the test, and without that, it was impossible to tell why the test failed.

ruffsl replied:
> I was unable to get the full console output when colcon ran the test

Could you be more specific? Do you have an example I could debug, or links to the two CI test runs I could compare?

crdelsey replied:
I just ran the system test locally using colcon. Here is the output:
```
> colcon test --merge-install --packages-select nav2_system_tests
Starting >>> nav2_system_tests
[Processing: nav2_system_tests]                   
[Processing: nav2_system_tests]                           
[Processing: nav2_system_tests]                             
[Processing: nav2_system_tests]                             
--- stderr: nav2_system_tests                               
Errors while running CTest
---
Finished <<< nav2_system_tests [2min 30s]	[ with test failures ]

Summary: 1 package finished [2min 30s]
  1 package had stderr output: nav2_system_tests
  1 package had test failures: nav2_system_tests

> 

> colcon test-result --verbose --test-result-base build/nav2_system_tests/
build/nav2_system_tests/Testing/20190814-2300/Test.xml: 11 tests, 0 errors, 1 failure, 0 skipped
build/nav2_system_tests/test_results/nav2_system_tests/test_localization.xml: 1 test, 0 errors, 1 failure, 0 skipped
- nav2_system_tests test_localization.missing_result
  <<< failure message
    The test did not generate a result file.
  >>>

Summary: 77 tests, 0 errors, 2 failures, 0 skipped

```

That's all I get - "test did not generate a result file"

Compare with any current build where I get all the console output - showing all the messages each node prints out.

ruffsl replied:
That summary is being generated with colcon by exampling what ctest wrote to [test_results/nav2_system_tests/test_localization.xml](https://528-166157381-gh.circle-artifacts.com/0/opt/overlay_ws/test_results/nav2_system_tests/test_localization.xml). 

I'm not sure why the ctest isn't reporting the stdout/stderr. My guess is that it wasn't configured to, but perhaps Dirk would know. Still, I don't think it would be useful to duplicate the stdout dump into the xml when its already captured from the log artfacts. I think that message field in the xml should be used for heigherlevel feedback reporting.

But looking at the colcon logs for the test, all the console output - showing all the messages each node prints out - seems to be there: logs/test/nav2_system_tests/... Please make sure that you can navigatiate the artifact tab for test jobs where this is all stored for you.

https://528-166157381-gh.circle-artifacts.com/0/opt/overlay_ws/log/test/nav2_system_tests/stdout.log
https://528-166157381-gh.circle-artifacts.com/0/opt/overlay_ws/log/test/nav2_system_tests/stdout_stderr.log
https://528-166157381-gh.circle-artifacts.com/0/opt/overlay_ws/log/test/nav2_system_tests/streams.log
https://528-166157381-gh.circle-artifacts.com/0/opt/overlay_ws/log/test/nav2_system_tests/command.log

![image](https://user-images.githubusercontent.com/2293573/63077065-edef8480-beeb-11e9-9795-4cbe5f861e3e.png)

crdelsey commented:
> Could I get an opinion on the test failures here

Those failures look expected except for this in the RTI connext test.
```
nav2_util test_actions.gtest.missing_result
  <<< failure message
    The test did not generate a result file:
    
    Running main() from /opt/ros/dashing/src/gtest_vendor/src/gtest_main.cc
    [==========] Running 2 tests from 1 test case.
    [----------] Global test environment set-up.
    [----------] 2 tests from ActionTest
    [ RUN      ] ActionTest.test_simple_action
    
    >>> [rcutils|error_handling.c:107] rcutils_set_error_state()
    This error state is being overwritten:
    
      'array_list is already initialized, at /home/jenkins-agent/workspace/packaging_linux/ws/src/ros2/rcutils/src/array_list.c:61, at /home/jenkins-agent/workspace/packaging_linux/ws/src/ros2/rcl/rcl/src/rcl/logging_rosout.c:179'
    
    with this new error message:
    
      'Failed to add publisher to map., at /home/jenkins-agent/workspace/packaging_linux/ws/src/ros2/rcl/rcl/src/rcl/logging_rosout.c:181'
    
    rcutils_reset_error() should be called after error handling to avoid this.
    <<<
    RTI Data Distribution Service Non-commercial license is for academic, research, evaluation and personal use only. USE FOR COMMERCIAL PURPOSES IS PROHIBITED. See RTI_LICENSE.TXT for terms. Download free tools at rti.com/ncl. License issued to Non-Commercial User license@rti.com For non-production use only.
    Expires on 00-jan-00 See www.rti.com for more information.
    RTI Data Distribution Service Non-commercial license is for academic, research, evaluation and personal use only. USE FOR COMMERCIAL PURPOSES IS PROHIBITED. See RTI_LICENSE.TXT for terms. Download free tools at rti.com/ncl. License issued to Non-Commercial User license@rti.com For non-production use only.
    Expires on 00-jan-00 See www.rti.com for more information.
    unknown file: Failure
    C++ exception with description "failed to initialize rcl node: Failed to add publisher to map., at /home/jenkins-agent/workspace/packaging_linux/ws/src/ros2/rcl/rcl/src/rcl/logging_rosout.c:181" thrown in SetUp().
  >>>
```

ruffsl commented:
I simplified the build config a bit using colcon mixin and repurposed the CI images into a debug and release for the different job executors. The ccache hit for debug with coverage enabled is still too low due to some nondeterminism I'm unsure of, so I just omitted ccache for the debug build.

mkhansen-intel commented:
@ruffsl - can you rebase and re-run the CI to get it passing? I'm assuming there's no point in reviewing this until the CI is passing

ruffsl commented:
I've been testing on my fork by merging this branch into upstream master on my own master branch to avoid breaking the commit links in the comments here with a force push. I've rebased and pushed with lease to this PR branch.

Looks like `debug_test` job is a little flaky, and `test_rmw_connext_cpp` has never passed, but the release tests for FastRTPS and Opensplice look healthy.

mkhansen-intel commented:
If test_rmw_connext_cpp doesn't work, let's not include it. I'd like to get all checks passing.

ruffsl commented:
The job is working, it's just not passing. Do we really want to disable any test that doesn't pass? You can see from the checks the passing of each job separately on GitHub, so I'm not sure hiding it to keep everything remaining green is in the spirit of testing all rmw.

mkhansen-intel commented:
No sense in adding a test that never passes. The fail would need to be debugged and fixed before we enable it.

ruffsl commented:
> No sense in adding a test that never passes.

But disabling a job is not the same as disabling a test. There about 908 total tests now, where `test_rmw_connext_cpp` job is already passing almost all of them. If you disable an rmw job, then commits/regressions that could increase number of failures would remain undiscoverable.

``` bash
API_URL=https://circleci.com/api/v1.1/project
VCS_TYPE=github
VCS_USERNAME=ruffsl
VCS_PROJECT=navigation2

function summarize_test_results { \
    curl -s $API_URL/$VCS_TYPE/$VCS_USERNAME/$VCS_PROJECT/$1/tests > \
        tests.json; \
    cat tests.json | \
        jq '.tests[] | .result' | \
        sort | uniq -c; \
}
```

### debug_test 

* https://circleci.com/gh/ruffsl/navigation2/496
* https://circleci.com/gh/ruffsl/navigation2/498

```
$ summarize_test_results 496
      1 "failure"
    907 "success"
$ summarize_test_results 498
    908 "success"
```

### test_rmw_connext_cpp

* https://circleci.com/gh/ruffsl/navigation2/494
* https://circleci.com/gh/ruffsl/navigation2/497

```
$ summarize_test_results 494
      1 "failure"
    907 "success"
$ summarize_test_results 497
      3 "failure"
    904 "success"
```

I'd suggest keeping it enabled on master at least to track the health history of each rmw implementations over time for triage. Committers can cancel PR jobs they don't need to consider, but given the testing jobs are all in parallel, it's not much of a bottleneck.

crdelsey commented on .dockerhub/debug/hooks/build:
```diff
@@ -0,0 +1,12 @@
+FROM_IMAGE=osrf/ros2:nightly
+FAIL_ON_BUILD_FAILURE=""
+UNDERLAY_MIXINS="debug ccache"
+OVERLAY_MIXINS="debug ccache coverage-gcc"
```
This mixin might need some additional work before we can use it. I can't remember the exact problem, but I needed to add the --coverage option to the linker commands as well. 
https://github.com/ros-planning/navigation2/blob/59f0360827a504ac9d54dfbc7019cd8547fc766c/nav2_common/cmake/nav2_package.cmake#L42-L43

I don't remember the exact problem though. I feel like it just left out coverage data rather than giving a linker error.

ruffsl replied:
Please review this related PR:
https://github.com/colcon/colcon-mixin-repository/pull/20

ruffsl replied:
@crdelsey , looks like the above PR has merged.

crdelsey replied:
I just confirmed the behavior. If I don't add the linker flags, some files just silently don't report coverage data. `install/include/costmap_queue\costmap_queue.hpp` is a good example. If I use the mixin, this file reports 0% coverage. If I add the linker flags, I get 67% coverage.

I don't fully understand why this is happening, though.

crdelsey replied:
I think the PR is good. I believe using --coverage allows compatibility with Clang. I'm not sure if it fixes this problem though. I'll try to do some tests today to figure out what's going on.

ruffsl replied:
Make sure you `colcon mixin update <name>` for the name you used when adding the default mixin-repository linked above, so that you have the latest changes.

crdelsey replied:
Yeah. It looks good now. I wonder why we didn't get linking errors before. But, I'm not curious enough to bother testing; I'm just happy that this works :-)

crdelsey commented:
> I'd suggest keeping it enabled on master at least to track the health history of each rmw implementations over time for triage

There are two problems with leaving it on:
1. People stop looking at the test results and just merge PRs.
2. The codecov.io report skips failing commits.

We can maybe address 1. by only running the connext test on the nightly build you set up. Maybe that would be good enough for 2 as well.

ruffsl commented on tools/code_coverage_report.bash:
```diff
@@ -15,7 +15,7 @@ set -e
 PWD=`pwd`
 
-COVERAGE_REPORT=genhtml
+COVERAGE_REPORT_VIEW="none"
```
There was already a `clean` command, so I assumed the intent was for users to be explicit in what they wanted the script to do. I've swapped the default and added a case for ci.

ruffsl commented on Dockerfile:
```diff
@@ -26,37 +40,49 @@ RUN . /opt/ros/$ROS_DISTRO/setup.sh && \
-ARG CMAKE_BUILD_TYPE=Release
-
+# build underlay source
+ARG UNDERLAY_MIXINS="release ccache"
```
The docker image is eventually being used to bootstrap the ccache for new PRs. Setting that mixin here allows for all the build effort that dockerhub upfronts to be cached across fresh workflows. Consecutive triggers for the same PR will restore their own ccache cache from circleci's workflow cache, to keep it warm just for that PR.

ruffsl commented:
> by only running the connext test on the nightly build you set up.

I can settle for that. I've moved the extra release tests to the nightly workflow so we can keep an eye on them, and remind us what's still failing, while keeping more container resource free for paralel PR jobs.

ruffsl commented:
Before merging, I'll have change the exciters over to use images from the `rosplanning/navigation2` DockerHub repo. I'm also still trying to get the dockerhub build triggers stable, so I'll post again when that's sorted.

crdelsey commented:
@ruffsl I want to add a doc build step to CI. It would do a Sphinx build of our doc folder (once the docs are checked in). It would then publish the output back to the gh-pages branch in this repo.

In a perfect world, we'd do the build part on each PR to ensure the docs build hasn't been broken. Then the build and publish would happen when master is updated.

Any recommendations on the best way to do this? As in, where should I fit this into the CircleCI config.

ruffsl commented:
> Any recommendations on the best way to do this? As in, where should I fit this into the CircleCI config.

Could you open a new issue describing what you had in mind, perhaps including any examples from other projects you'd like to replicate? Fitting this into the current CircleCi config doesn't seem hard.

I have a few questions what is the convention for documentation on ROS2; should this be hosted on github pages, or index.ros.org. How should deployment of docs be versioned, etc. We can expand a bit more under the ticket. I like the idea of offloading some of the hosting for releases to readthedocs.org if we can. Here is a related article on these topics:
https://circleci.com/blog/automate-releases-from-pipelines-using-infrastructure-as-code

ruffsl commented:
Ok, I've updated the automated build rules on docker hub and changed over the images in the config.
I'll merge and verify the link to osrf/ros2:nightly is triggering tomorrow.

ruffsl commented:
@crdelsey , could you go into the CircleCI security settings and enable 3rd party orbs:
https://circleci.com/gh/organizations/ros-planning/settings#security

![image](https://user-images.githubusercontent.com/2293573/63400011-40ff8680-c387-11e9-93db-3727fc05123c.png)

Otherwise we'll continue to receive CI errors like when we try and load the partnered codecov orb:
https://circleci.com/orbs/registry/orb/codecov/codecov

> Orb codecov/codecov@1.0.5 not loaded. To use this orb, an organization admin must opt-in to using third party orbs in Organization Security settings.

crdelsey commented:
@ruffsl I don't have access to that either. Before I ask OSRF to enable this for us, does using the ORB buy us anything over just using the upload script we used before?

ruffsl commented:
> does using the ORB buy us anything over just using the upload script we used before?

I'd evenly like to push more of the common commands into orbs of their own for working with ROS, so that we can share more CI boilerplate across ROS projects/community, and simplify config complexity. 
So to use our own self published orb, we'd still need the same setting.

crdelsey commented:
@tfoote Would you enable support for 3rd party orbs in the CircleCI settings for the ROS-Planning organization as @ruffsl describes [3 notes above](https://github.com/ros-planning/navigation2/pull/990#issuecomment-523278485)?

We're using CircleCI as the CI system for navigation2 and using CodeCov.io to provide a code coverage report. CodeCov.io provides a plugin (orb) to make it easier to integrate with CircleCI, but, to use it, we need to grant it permissions on the CircleCI website. This has to be done by someone with permissions at the github ros-planning org level, so I can't enable it myself

If you've never used CircleCI before, there's an option to "login with github" which will associate the CircleCI account with your Github account and grant you appropriate permissions.

Thanks

tfoote commented:
@nuclearsandwich Is going to enable it for you.

nuclearsandwich commented:
![wtfisorbs](https://user-images.githubusercontent.com/358882/63623295-392e2500-c5c7-11e9-9718-9589408c1f20.png)

In theory you're :heavy_check_mark:

crdelsey commented:
@nuclearsandwich Thanks. Looks good.
